src/swiss_dict.jl

```diff
@@ -0,0 +1,559 @@
+    end
+    return h
+end
+# Note the constructors of WeakKeyDict mirror these here, keep in sync.
+SwissDict() = SwissDict{Any,Any}()
+SwissDict(kv::Tuple{}) = SwissDict()
+copy(d::SwissDict) = SwissDict(d)
+empty(d::SwissDict, ::Type{K}, ::Type{V}) where {K, V} = SwissDict{K, V}()
+
+const AnyDict = SwissDict{Any,Any}
+
+SwissDict(ps::Pair{K,V}...) where {K,V} = SwissDict{K,V}(ps)
+SwissDict(ps::Pair...)                  = SwissDict(ps)
+
+function SwissDict(kv)
+    try
+        dict_with_eltype((K, V) -> SwissDict{K, V}, kv, eltype(kv))
+    catch
+        if !isiterable(typeof(kv)) || !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("SwissDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow()
+        end
+    end
+end
+
+function grow_to!(dest::AbstractDict{K, V}, itr) where V where K
+    y = iterate(itr)
+    y === nothing && return dest
+    ((k,v), st) = y
+    dest2 = empty(dest, typeof(k), typeof(v))
+    dest2[k] = v
+    grow_to!(dest2, itr, st)
+end
+
+# this is a special case due to (1) allowing both Pairs and Tuples as elements,
+# and (2) Pair being invariant. a bit annoying.
+function grow_to!(dest::AbstractDict{K,V}, itr, st) where V where K
+    y = iterate(itr, st)
+    while y !== nothing
+        (k,v), st = y
+        if isa(k,K) && isa(v,V)
+            dest[k] = v
+        else
+            new = empty(dest, promote_typejoin(K,typeof(k)), promote_typejoin(V,typeof(v)))
+            merge!(new, dest)
+            new[k] = v
+            return grow_to!(new, itr, st)
+        end
+        y = iterate(itr, st)
+    end
+    return dest
+end
+
+empty(a::AbstractDict, ::Type{K}, ::Type{V}) where {K, V} = SwissDict{K, V}()
+
+#hashindex(key, sz) = (((hash(key)%Int) & (sz-1)) + 1)::Int
+
+##Simd utilities
+@inline _expand16(u::UInt8) = ntuple(i->VecElement(u), Val(16))
+_blsr(i::UInt32)= i & (i-Int32(1))
+
+@inline _vcmp_eq(u::_u8x16, v::_u8x16) = Core.Intrinsics.llvmcall(("""
+%cmp = icmp eq <16 x i8> %0, %1
+%cmp16 = bitcast <16 x i1> %cmp to i16
+%res = zext i16 %cmp16 to i32
+ret i32 %res
+"""), UInt32, Tuple{_u8x16,_u8x16}, u, v)
+
+@inline _vcmp_le(u::_u8x16, v::_u8x16) = Core.Intrinsics.llvmcall(("""
+%cmp = icmp ule <16 x i8> %0, %1
+%cmp16 = bitcast <16 x i1> %cmp to i16
+%res = zext i16 %cmp16 to i32
+ret i32 %res
+"""), UInt32, Tuple{_u8x16,_u8x16}, u, v)
+
+@inline function _prefetchr(p::Ptr)
+    p = reinterpret(UInt, p)
+    if UInt === UInt64
+        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
+    %ptr = inttoptr i64 %0 to i8*
+    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    else
+        Core.Intrinsics.llvmcall(("declare void @llvm.prefetch(i8*, i32, i32, i32)", """
+    %ptr = inttoptr i32 %0 to i8*
+    call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
+    ret void
+    """), Nothing, Tuple{UInt}, p)
+    end
+end
```

Comment:
Can we get rid of the `if` by interpolating the type string in?

```suggestion
@inline function _prefetchr(p::Ptr)
    p = reinterpret(UInt, p)
    int_t = Int === Int64 ? "i64" : "int_t"
    Core.Intrinsics.llvmcall(
        ("declare void @llvm.prefetch(i8*, i32, i32, i32)",
        """
        %ptr = inttoptr $int_t %0 to i8*
        call void @llvm.prefetch(i8* %ptr, i32 0, i32 3, i32 1)
        ret void
        """),
        Nothing, Tuple{UInt}, p
    )
end
```